### DISCUSSION TITLE
Flutter phone authentication with play integrity api.

### DISCUSSION THREAD

**Discussion by salman0017** [firebase/flutterfire#10523] [Q&A]
Flutter phone authentication not working since safety net device verification is deprecated.

For the new projects created after January safety net option not available we must use play integrity api.

1. Enabled play integrity api from firebase project.
2. Linked Firebase project as suggested here https://firebase.google.com/docs/app-check/android/play-integrity-provider.
3. enabled play integrity from flutter code as suggested here https://firebase.google.com/docs/app-check/flutter/default-providers 

still its redirecting to browser page asking us to verify app with reCAPTCHA.

in the browser page im getting this error: _'Unable to process request due to missing initial state. This may happen if browser sessionStorage is inaccessible or accidentally cleared.'_

when we press back I'm getting  'Missing client identifier error'

**Comment by Technorocker** [comment]
I too am getting this error and cant figure out why.

I have setup Play integrity in google cloud and link myproject, also using the debug token thats given from the app and registered it inside the project.

**Reply by harshanacz** [reply]
**Did you find any solution for it?**

**Comment by luckyhandler** [comment]
There's an ongoing incident on this topic
https://status.firebase.google.com/incidents/UY1LTFan8X5oybhphzfV

Does this mean that phone authentication on Android is currently not working? If yes, I don't understand why this is not considered high priority. The `Device Verification API` cannot be activated any longer and the new setup doesn't work. Or are we missing anything?

**Reply by luckyhandler** [reply]
I had to remove the "Android apps" restriction from my Android API key --> now at least the reCAPTCHA flow works!

**Reply by luckyhandler** [reply]
As of today the Play Integrity API is working again and the app doesn't go into the reCAPTCHA flow any more but uses the primary one.

**Reply by Daquiver1** [reply]
Can you show me how you implemented it? I've followed the docs and made all the changes but whenever a user logs in, it initiates the web recaptcha flow. Spent hours on it and don't know what to do right now.

**Comment by nohe427** [comment]
Thanks for reporting this issue. I believe that there are potentially a few instances where the fallback to reCAPTCHA could occur:
1. The wrong version of firebase\_auth is used. firebase\_auth version 4.4.0 is the first version to support Play Integrity API for SMS verification.
2. You only have a SHA-1 certificate fingerprint registered in the Firebase console. You will need to have the SHA-256 fingerprint certificate registered.
3. You have the wrong certificate registered. You may have accidentally used your upload certificate rather than the Google Play Signing certificate. You can validate that the Google Play Signing certificate is registered by visiting the Google Play Console, selecting your application, and then in the right hand bar, under Setup, selecting App Integrity. Then in the main window, select App signing and validate that your App signing key certificate (SHA-256 and SHA-1) are registered in the Firebase console. You can also use [Peter’s Asset Link Tool](https://goo.gle/3oqWmip) to validate that the app deployed on your device is using the right fingerprint as well. [This video](https://youtu.be/3bAQPnxLd4c?t=120), while a little outdated and targeted for a web audience, can help you find where in the console to grab your SHA-256 fingerprint.
4. You deployed straight to your phone rather than downloaded the app from the Play Store. The app needs to come from the Play Store on a real device for Play Integrity API to be used. Otherwise, it may fall back to reCAPTCHA.
5. You may have exhausted your [daily quota](https://console.cloud.google.com/apis/api/playintegrity.googleapis.com/quotas?project=_) of 10,000 API calls per day. If so, you can [request a lift on the limit](https://support.google.com/googleplay/android-developer/answer/11395166?hl=en-GB#:~:text=Monitor%20your%20Play%20Integrity%20API,day%20to%20the%20Integrity%20API.).


A few other notes:
*   **App Check is not required to be set up on the device**. Play Integrity API is baked into the firebase\_auth package and no additional client setup is required (other than registering certificate fingerprints from the Google Play Console, etc..).
*   Off play flows are not supported - you must download the app from Google Play and be running on a real device for this to work.
*   You must register the Google Play Signing Certificate if your app uses Play Signing (unless you were published prior to August 2021 and didn’t upgrade to Play Signing, you likely are using Play Signing).
*   Also, ensure you are running the latest version of Google Play services.

@luckyhandler - Phone authentication is still working on Flutter Android and Android. Phone authentication is falling back to reCaptcha for attestation rather than using the Play Integrity API.

If you are unable to get the reCaptcha flow working as well, visit [this comment](https://github.com/firebase/flutterfire/issues/10593#issuecomment-1516862489) which provides some additional troubleshooting and validation steps to ensure that your app is able to at least fallback to reCaptcha while Play Integrity is being restored as the client verification method.

**Reply by luckyhandler** [reply]
@nohe427 I checked the keys probably 10 times by now. All are correctly set and it previously also worked with the Device Verification Api

**Reply by nohe427** [reply]
@luckyhandler -
* Which version of [Firebase Auth](https://pub.dev/packages/firebase_auth) are you using?
* Can you create a minimum reproducible sample?
* Is there a specific error message that you are getting when running through the flow?
* Are you using Firebase Auth UI or something else?
* Does the reCaptcha fail to load or appear for you or does the sms message not get delivered?
* What device are you using for testing?

**Reply by luckyhandler** [reply]
ok, thanks for your help @nohe427 . Removing the "Android Apps" restriction from my Android API key solved the problem with the reCAPTCHA flow. So users can log in again with with the fallback flow until the Integrity API issue is solved!

**Reply by ridwan-aswan** [reply]
> oke, terima kasih atas bantuanmu@nohe427. Menghapus batasan "Aplikasi Android" dari kunci API Android saya memecahkan masalah dengan alur reCAPTCHA. Jadi pengguna dapat masuk lagi dengan alur fallback hingga masalah Integrity API terungkap!

I have done that, but the sms from firebase is not the application name, but like this, Enter: XXXXXX App: appname.firebaseapp.com , the sms is like in debug mode.

**Reply by nohe427** [reply]
That is an expected behavior at the moment since [Play Integrity is not serving client verifications requests](https://status.firebase.google.com/incidents/UY1LTFan8X5oybhphzfV) at the moment. Once Play Integrity is reactivated, it should return to the previous SMS message.

**Reply by AndrewTran2018** [reply]
> ok, thanks for your help @nohe427 . Removing the "Android Apps" restriction from my Android API key solved the problem with the reCAPTCHA flow. So users can log in again with with the fallback flow until the Integrity API issue is solved!

@luckyhandler what is  Android API key?

**Reply by luckyhandler** [reply]
Hey @AndrewTran2018 It's the key you use in your google-services.json file and which is defined under `api_key`. You can find it at https://console.cloud.google.com/apis/credentials under `API Keys`

**Reply by AndrewTran2018** [reply]
> Hey @AndrewTran2018 It's the key you use in your google-services.json file and which is defined under `api_key`. You can find it at https://console.cloud.google.com/apis/credentials under `API Keys`

Thank, I saw it. Do I need to activate to begin integrating PlayIntegrity with 300 bucks credit? @luckyhandler

**Reply by nohe427** [reply]
> > Hey @AndrewTran2018 It's the key you use in your google-services.json file and which is defined under `api_key`. You can find it at https://console.cloud.google.com/apis/credentials under `API Keys`
> 
> Thank, I saw it. Do I need to activate to begin integrating PlayIntegrity with 300 bucks credit? @luckyhandler 


If you are still seeing the reCaptcha page, it's due to play integrity not being set as the client verification method. This is expected behavior at the moment due to https://status.firebase.google.com/incidents/UY1LTFan8X5oybhphzfV and potentially not having a Google play deployed app.

Removing api key restrictions is for if reCaptcha appears but no sms message is sent. Are you still getting sms messages?

**Reply by AndrewTran2018** [reply]
> > > Hey @AndrewTran2018 It's the key you use in your google-services.json file and which is defined under `api_key`. You can find it at https://console.cloud.google.com/apis/credentials under `API Keys`
> > 
> > 
> > Thank, I saw it. Do I need to activate to begin integrating PlayIntegrity with 300 bucks credit? @luckyhandler
> 
> If you are still seeing the reCaptcha page, it's due to play integrity not being set as the client verification method. This is expected behavior at the moment due to https://status.firebase.google.com/incidents/UY1LTFan8X5oybhphzfV and potentially not having a Google play deployed app.
> 
> Removing api key restrictions is for if reCaptcha appears but no sms message is sent. Are you still getting sms messages?

Yes, I still receive sms messages on device. However, it took so long.

**Reply by AndrewTran2018** [reply]
> > > Hey @AndrewTran2018 It's the key you use in your google-services.json file and which is defined under `api_key`. You can find it at https://console.cloud.google.com/apis/credentials under `API Keys`
> > 
> > 
> > Thank, I saw it. Do I need to activate to begin integrating PlayIntegrity with 300 bucks credit? @luckyhandler
> 
> If you are still seeing the reCaptcha page, it's due to play integrity not being set as the client verification method. This is expected behavior at the moment due to https://status.firebase.google.com/incidents/UY1LTFan8X5oybhphzfV and potentially not having a Google play deployed app.
> 
> Removing api key restrictions is for if reCaptcha appears but no sms message is sent. Are you still getting sms messages?

And yes, the reCatcha page appeared @nohe427

**Comment by AndrewTran2018** [comment]
It is clearly that at present Play Integrity migration is a big hurt for developers, so the deadline of Play Store is ridiculous.

**Reply by AndrewTran2018** [reply]
> It is clearly that at present Play Integrity migration is a big hurt for developers, so the deadline of Play Store is ridiculous.

@nohe427 
Do you have any suggestion? After several times of login on device, I received device blocking messages from Google. Really tired.

**Reply by AndrewTran2018** [reply]
@luckyhandler Should I roll back to using SafetyNet until the problem is fixed?

**Reply by luckyhandler** [reply]
@AndrewTran2018 I don't know what you mean by rolling back, but if you deactivated the "Device Verification API" - like me - it's a point of no return. You will have to wait until the bug is fixed.

**Reply by AndrewTran2018** [reply]
> Device Verification API"

I only found "[Google Play Integrity API]" in GCC, is that a replacement of "Device Verification API" being deprecated?
@luckyhandler

**Comment by Patryk999** [comment]
I have problem related to this issue. Captcha flow works locally but it looks like it doesn't work on Google Play app version.
Followed all steps listed by @nohe427, configured SHA-1 properly and doesn't have restriction for Android API key.
It looks like the app is not even opening the browser.
Do you have any suggestion what might be missing?

**Reply by nohe427** [reply]
Do you have the sha-256 cert fingerprint from the play console added to firebase?

**Reply by Patryk999** [reply]
> Do you have the sha-256 cert fingerprint from the play console added to firebase?

Yes, both sha-256 and sha-1.

**Reply by AndrewTran2018** [reply]
> I have problem related to this issue. Captcha flow works locally but it looks like it doesn't work on Google Play app version. Followed all steps listed by @nohe427, configured SHA-1 properly and doesn't have restriction for Android API key. It looks like the app is not even opening the browser. Do you have any suggestion what might be missing?

Mine exactly the same, even only AndroidProvider.debug works on emulator. Following ...

**Reply by nohe427** [reply]
> > Do you have the sha-256 cert fingerprint from the play console added to firebase?
> 
> Yes, both sha-256 and sha-1.

What does log cat give you for information? What version of firebase_auth and firebase_core are running on the Play Store version?

**Comment by AndrewTran2018** [comment]
@nohe427 @luckyhandler Please help check the steps I did to integrate Play Integrity:
1. Enable Play Integrity API in GCC, link the project in Play Store Console, 
2. Registering it in Firebase Console. 
3. Add code in [the guide](https://firebase.google.com/docs/app-check/flutter/default-providers) and now only AndroidProvider.debug works with reCaptchar fallback on emulator.

Many thanks.

**Reply by AndrewTran2018** [reply]
Halo @nohe427,

It falls back to reCaptCha flow on device now, but does not send sms.

**Reply by nohe427** [reply]
> 3\. the guide

AppCheck and Auth are two different things. The guide you sent is for AppCheck, so that may be where you are getting stuck.

Based on only getting reCaptcha flow working on emulators, it sounds like you did not register the Play Store SHA 256 and SHA 1 keys,  but may have only registered your SHA256 key from the Play Store. This guide should help with validating your setup : https://github.com/firebase/flutterfire/issues/10593#issuecomment-1516862489

**Reply by AndrewTran2018** [reply]
> > 3. the guide
> 
> AppCheck and Auth are two different things. The guide you sent is for AppCheck, so that may be where you are getting stuck.
> 
> Based on only getting reCaptcha flow working on emulators, it sounds like you did not register the Play Store SHA 256 and SHA 1 keys, but may have only registered your SHA256 key from the Play Store. This guide should help with validating your setup : [#10593 (comment)](https://github.com/firebase/flutterfire/issues/10593#issuecomment-1516862489)

Nope! I did all of these things already. I did not mention the configuration in the prev post because it is de facto. Everything is correct.  This morning, the SMS message finally arrived my phone after almost an hour., and of course expired already. Check out the attached screen. It is the problem of Google and many developers have to suffer. @nohe427 
![Screenshot_20230428_133406](https://user-images.githubusercontent.com/31315360/235074416-e9592b4e-17b4-4c76-9413-1fefe309225f.jpg)

**Reply by nohe427** [reply]
@AndrewTran2018 - Which region are you in and which region is your project located in? Can you confirm that you have tried the SMS OTP flow in a different physical location to ensure its not due to cell reception?

**Reply by AndrewTran2018** [reply]
This morning, first time ever, the OTP sms came in time and the authentication process ran as designed. It seems the problem gets fixed. @nohe427

**Reply by nohe427** [reply]
> This morning, first time ever, the OTP sms came in time and the authentication process ran as designed. It seems the problem gets fixed. @nohe427 

Yes, I posted an update in the other thread
 Glad to hear it's working for you now

**Comment by liuzs0666** [comment]
We get impacted by this bug. Double checked the SHA256 and SHA1 key matches my local APK and the one on Play Store console.

More importantly, our clients have reported that reCaptcha also fails for them. In the popup web page, it says:
> Unable to process request due to missing initial state. This may happen if browser sessionStorage is inaccessible or accidentally cleared.

Please help! This is blocking the normal user flow.

**Comment by Daquiver1** [comment]
From reading the issue, I'm assuming that play integrity requires the app to be on the play store, but If that's the case then how can I use it if I'm still in the production stage?
 From the flutter firebase docs, they suggested that I use AndroidProvider.Debug for the development stage, but when I do that and turn on app check enforcement, all my requests get blocked with the message Invalid token.
And I've confirmed that I'm using the right sha-256.

**Reply by nohe427** [reply]
For pre-production, the recommendation is to [use fictional numbers](https://firebase.google.com/docs/auth/android/phone-auth#test-with-fictional-phone-numbers). The AndroidProvider.Debug is for AppCheck, not for Phone Authentication.

**Comment by EngrTAHAR-Noureddine** [comment]
Hello everyone! I wanted to share the solution I found for a problem I was facing. Initially, I hadn't restricted the API key for Google Cloud of Firebase, specifically the Android API key. I did this to test if Firebase Auth was functioning correctly. After confirming that Firebase phone authentication was working fine, I proceeded to restrict the API key by using the app signing key from the Google Console and the SHA-1 from my Android Studio for debugging purposes. Hence, the issue I was encountering did not relate to AppCheck or activating Android device verification. The key point was to ensure that the Firebase API key was restricted in a manner similar to the API key used by Google Cloud. I'm grateful for this forum, which provided me with the necessary solutions and hints. Thank you!
